### PR TITLE
User log spec

### PR DESCRIPTION
**Pull request by tonytamps**
Addresses issue #61 

Includes the ability to filter and tag the log messages.

### CODE REVIEW

**Comment by thomwiggers** [comment]
This is somewhat related to #8 and has some overlap...

**Comment by ghost** [comment]
@tonytamps So here is a general overview of what this is.

Every user has a log associated with them, this log serves as a paper trail for certain actions. For example if a user gets ip banned due to incorrect login attempts a record is written into the user's log indicating this. Furthermore the user log contains information inputted manually by moderators. For example if a user has accumulated forum warnings and had to be talked with, a note would be left in the log with information.

As it stands in the current php version of gazelle the user log is just a blob of text (I know, a very questionable decision by the original developers). We can't make those kind of mistakes. I envision entries in the user log would have categories and sub categories associated with them in order to make it manageable and searchable. Furthermore the usage of specific categories should be restricted based on permissions. For example, forum mods should be able to leave forum related entries. 

I posted in #138 about how a generic tagging system would be useful, this is another potential use case for it.  

Also something to consider in the back of your mind is how we'll migrate data into this new database. A discussion worthy of a separate issue. Something I was thinking is simply maintaining a legacy log rather than trying to migrate it....

**Comment by tonytamps** [comment]
@ajax-wcd ok, I follow. So it won't be an audit log for all actions like the spec I wrote implies? Should it be more like that? I think it would be potentially useful to the user.

Having the admins manually add entries into the log makes it seem like the user shouldn't actually see it?

**Comment by ghost** [comment]
@tonytamps The the log should be permission based, so certain entries are only be visible when a user's permissions requirements are satisfied.

**Comment by tonytamps** [comment]
This is still being worked on, I'm just slow. Only some of the changes have been made so far.

**Comment by tonytamps** [comment]
@ajax-wcd updated to include tagging

File: doc/specs/userLog.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+* No messages outside of the user's permissions were shown.
+
+**Basic Flow:**
+
+1. The user navigates to their user log page.
+2. The user log messages are displayed.
+
+#### 1.2 As an administrator
+
+**Primary Actor:** An administrator.
```
Comment (ghost) [inline doc/specs/userLog.md]: Don't use role-specific names (Administrator, Moderator, etc.)

Instead, use something like "A user with the user log administration permission". This applies in several cases.

File: doc/specs/userLog.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+**Brief:** The user log is visible to the administrator.
+
+**Preconditions:**
+* The user log contains messages with varying tags.
```
Comment (ghost) [inline doc/specs/userLog.md:52]: I don't think this is a valid precondition. If the log was empty, the page would just show nothing, or a message like "No log entries for user found".

File: doc/specs/userLog.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+**Postconditions:**
+* No messages outside of the administrator's permissions were shown.
+
+**Basic Flow:**
+
+1. The administrator navigates to the user's log page.
+2. The user log messages are displayed.
+
+**Non-Functional Requirements:** N/A
```
Comment (ghost) [inline doc/specs/userLog.md]: Omit sections that don't apply.

File: doc/specs/userLog.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+### 2.0 Filtering the log
+
+**Primary Actor:** A user.
+
+**Brief:** The filtered results of the user log are filtered correctly.
+
+**Preconditions:**
+* The user log contains messages with varying tags.
```
Comment (ghost) [inline doc/specs/userLog.md:74]: ~~See above regarding this precondition.~~

Edit: In this use case, this precondition makes sense. Ignore this.

File: doc/specs/userLog.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+**Postconditions:**
+* No messages outside of the user's permissions were shown.
+
+**Basic Flow:**
+
+1. The user navigates to their user log page.
+2. The filter is set to the desired day.
+3. Filtered messages from the log are displayed.
+
+**Non-Functional Requirements:** N/A
```
Comment (ghost) [inline doc/specs/userLog.md]: See above.

File: doc/specs/userLog.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+**Brief:** Messages can be added to the log by an administrator.
+
+**Preconditions:**
+* The administrator has the required permissions to add messages to a user's log.
+
+**Trigger:** Administrator wants to add a message into a user's log.
+
+**Postconditions:**
+* Message was able to
```
Comment (ghost) [inline doc/specs/userLog.md]: Message was able to what?

File: doc/specs/userLog.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+**Basic Flow:**
+
+1. The administrator navigates to the user's log page.
+2. The administrator adds a message.
+3. The administrator selects a tag.
+4. The message is recorded.
+
+**Non-Functional Requirements:** N/A
```
Comment (ghost) [inline doc/specs/userLog.md]: See above.

File: doc/specs/userLog.md (hunk 2 above)
Comment (tonytamps) [inline doc/specs/userLog.md:52]: I mostly agree with you @guitarman0831 but I think that should be a separate use-case.

If no messages with tags exist, the assertion at the end that "only messages visible to the user's permission are displayed" is unable to be verified.

I think this precondition should stay here and another case created for "when no messages exist, no messages are shown"

Thoughts?

File: doc/specs/userLog.md (hunk 1 above)
Comment (tonytamps) [inline doc/specs/userLog.md]: Resolved.

File: doc/specs/userLog.md (hunk 3 above)
Comment (tonytamps) [inline doc/specs/userLog.md]: resolved

File: doc/specs/userLog.md (hunk 6 above)
Comment (tonytamps) [inline doc/specs/userLog.md]: ...get distracted?

Resolved.

File: doc/specs/userLog.md (hunk 7 above)
Comment (tonytamps) [inline doc/specs/userLog.md]: resolved.

File: doc/specs/userLog.md (hunk 2 above)
Comment (tonytamps) [inline doc/specs/userLog.md:52]: Resolved

**Comment by tonytamps** [comment]
I think this is done now.

**Pull request merged**